File: Labs/lab6/Lab6.py
'''
Created on Mar 27, 2024

@author: alexxavier
'''
from collections import deque

class Lab6:
    
    def canFinish(self, numCourses, prerequisites):
        graph = [[] for _ in range(numCourses)]
        
        for course, prereq in prerequisites:
            graph[course].append(prereq)
        
        processed = 0
        queue_of_courses = deque()
        
        for i in range(numCourses):
            if not graph[i]:
                queue_of_courses.append(i)
        
        while queue_of_courses:
            curr = queue_of_courses.popleft()
            processed += 1
            
            for i in range(numCourses):
                if curr in graph[i]:
                    graph[i].remove(curr)
                    if not graph[i]:
                        queue_of_courses.append(i)
        
        return processed == numCourses


lab = Lab6()
array1 = [[1, 0], [0, 1]]
array2 = [[1, 0]]

print(lab.canFinish(3, array1))  
print(lab.canFinish(3, array2)) 

        
        
    

    

File: Labs/lab6/Lab6.java
package Labs;

import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedQueue;

public class Lab6 {

	public static boolean canFinish(int numCourses, int[][] prerequisites) {
		List<List<Integer>> graph = new ArrayList<List<Integer>>();
		for (int i = 0; i < numCourses; i++) {
			graph.add(i, new ArrayList<Integer>());
		}

		for (int i = 0; i < prerequisites.length; i++) {

			if (prerequisites[i][0] == prerequisites[i][1]) {
				return false;
			}

			graph.get(prerequisites[i][0]).add(prerequisites[i][1]);
		}

		Queue<Integer> queueOfCourses = new ConcurrentLinkedQueue<Integer>();
		int processed = 0;

		while (processed != numCourses) {
			for (int i = 0; i < graph.size(); i++) {
				if (graph.get(i).size() == 0) {
					queueOfCourses.add(i);
				}
			}

			if (queueOfCourses.isEmpty()) {
				return false;
			}

			while (!queueOfCourses.isEmpty()) {
				Integer curr = queueOfCourses.poll();
				graph.get(curr).add(numCourses + 1);
				processed++;

				for (int i = 0; i < graph.size(); i++) {
					for (int j = 0; j < graph.get(i).size(); j++) {
						if (graph.get(i).get(j) == curr) {
							graph.get(i).remove(j);
						}
					}

				}
			}
		}
		return true;
	}

	public static void main(String[] args) {
		int[][] array1 = { { 1, 0 }, { 0, 1 } };
		int[][] array2 = { { 1, 0 } };

		System.out.println(canFinish(3, array1));
		System.out.println(canFinish(3, array2));

	}
}
